src/Rendering/VTKJS/Images/applyRenderedImage.js

```diff
@@ -244,86 +281,78 @@ function applyRenderedImage(context, { data: { name } }) {
-              break
-            case 'Float32Array':
-              range[0] = 0.0
-              range[1] = 1.0
-              break
-            case 'Float64Array':
-              range[0] = 0.0
-              range[1] = 1.0
-              break
-            default:
-              console.error('Unsupported data type')
-          }
-        }
-        if (!actorContext.colorRangeBounds.has(componentIndex)) {
+  representationProxy.getMapper().setMaximumSamplesPerRay(2814)
+
+  // update color ranges
+  actorContext.visualizedComponents
+    .map((componentIndex, fusedImageIndex) => [componentIndex, fusedImageIndex])
+    .filter(([componentIndex]) => componentIndex >= 0) // skip label map
+    .forEach(([componentIndex, fusedImageIndex]) => {
+      const {
+        colorRangeBoundsTouched,
+        colorRangeBounds,
+        colorRanges,
+        colorRangesTouched,
+      } = actorContext
+
+      const dataArray = actorContext.fusedImage.getPointData().getScalars()
+      const [dataMin, dataMax] = dataArray.getRange(fusedImageIndex)
+
+      const [newMin, newMax] =
+        !actorContext.independentComponents || dataMin - dataMax === 0
+          ? getDefaultRangeByDataType(dataArray.getDataType())
+          : [dataMin, dataMax]
+
+      if (!colorRangeBoundsTouched.get(componentIndex)) {
+        const oldRange = colorRangeBounds.get(componentIndex) ?? [
+          Number.POSITIVE_INFINITY,
+          Number.NEGATIVE_INFINITY,
+        ]
+        // only grow range
+        const newRange = [
+          Math.min(newMin, oldRange[0]),
+          Math.max(newMax, oldRange[1]),
+        ]
+        const hasChanged = newRange.some((value, i) => value !== oldRange[i])
+        if (hasChanged) {
           context.service.send({
             type: 'IMAGE_COLOR_RANGE_BOUNDS_CHANGED',
-            data: { name, component: componentIndex, range },
+            data: {
+              name,
+              component: componentIndex,
+              range: newRange,
+              fromAutomation: true,
+            },
           })
         }
-        if (!actorContext.colorRanges.has(componentIndex)) {
+      }
+
+      const storedColorRange = colorRanges.get(componentIndex)
+      const oldRange = storedColorRange ?? [
+        Number.POSITIVE_INFINITY,
+        Number.NEGATIVE_INFINITY,
+      ]
+      if (!colorRangesTouched.get(componentIndex) || !storedColorRange) {
+        // only grow range
+        const newRange = [
+          Math.min(newMin, oldRange[0]),
+          Math.max(newMax, oldRange[1]),
+        ]
+        const hasChanged = newRange.some((value, i) => value !== oldRange[i])
+
+        if (hasChanged) {
           context.service.send({
             type: 'IMAGE_COLOR_RANGE_CHANGED',
-            data: { name, component: componentIndex, range },
+            data: {
+              name,
+              component: componentIndex,
+              range: newRange,
+              fromAutomation: true,
```

Comment:
fromROI ?

Reply:
Now `keepAutoAdjusting`.  Yes, this grow bounds if user has not messed with them logic is confusing.  Not sure how to improve...